In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


### Get the number of unique streams per organizations from the stream_details csv, which captures organization, stream, card mapping

In [7]:
stream_details_df = pd.read_csv("../data/stream_details.csv", encoding = "ISO-8859-1")
print(stream_details_df.head())

   ORGANIZATION  STREAMID                STREAMNAME  CARDID  \
0            14       163  Lubes Concepts for Japan    1090   
1            14       163  Lubes Concepts for Japan    1091   
2            14       163  Lubes Concepts for Japan    1092   
3            14       163  Lubes Concepts for Japan    1093   
4            14       163  Lubes Concepts for Japan    1094   

                 CARDNAME  
0            ????????????  
1    ???????????? - ?????  
2    ???????????? - ?????  
3  ???????????? - ???????  
4    ???????????? - ?????  


### Get the number of unique organizations

In [8]:
print(stream_details_df["ORGANIZATION"].nunique())

2


### Get the number of unique streams

In [9]:
print(stream_details_df["STREAMID"].nunique())

840


### Get the number of cards per stream

In [33]:
cards_per_stream = stream_details_df.groupby('STREAMID')['CARDID'].nunique()
print(cards_per_stream)

STREAMID
163     15
167     14
171      6
172      4
173      8
186      6
199     12
201     10
202      1
203     15
204      1
205      8
206      9
218     21
219      1
228      4
229      1
230      1
231      8
232      8
233      6
236      1
237      1
244     18
245      7
246      8
247      1
248      3
315      3
316      1
        ..
1657    14
1659     2
1660    27
1661     1
1662     6
1663     7
1664     7
1665     4
1666     3
1668     2
1669     5
1670     8
1671     7
1672     3
1673     2
1674    17
1677     1
1678     1
1680     7
1682     1
1683     7
1684     2
1685     1
1686     1
1687     1
1688     1
1689     1
1690     1
1691     1
1694     2
Name: CARDID, Length: 840, dtype: int64


### Get the distribution of streams for each organization - group by the organization and then the number of streams

In [11]:
stream_details_by_org = stream_details_df.groupby('ORGANIZATION')['STREAMID'].nunique()
print(stream_details_by_org)

ORGANIZATION
14    120
28    720
Name: STREAMID, dtype: int64


In [12]:
users_details_df = pd.read_csv("../data/UserProfile_Customer1 and 2.csv", encoding = "ISO-8859-1")
print(users_details_df.head())

   org_id  User_Id  Gender  Dob     Location    City Country  \
0      14      245     NaN  NaN          NaN     NaN      SG   
1      14      337     NaN  NaN          NaN     NaN      SG   
2      14      267     NaN  NaN  Philippines  Manila      PH   
3      14      343     NaN  NaN          NaN     NaN      SG   
4      14      574     NaN  NaN          NaN     NaN      PH   

               Role user_timezone  
0  No Role Assigned     Singapore  
1  No Role Assigned     Singapore  
2  No Role Assigned     Singapore  
3  No Role Assigned     Singapore  
4  No Role Assigned     Singapore  


### Get the number of unique organizations from the user table

In [13]:
print(users_details_df["org_id"].nunique())

2


### Get the number of unique users from the user table

In [15]:
print(users_details_df["User_Id"].nunique())

22412


### Get the distribution of users for each organization - group by the organization and then the number of users

In [16]:
user_details_by_org = users_details_df.groupby('org_id')['User_Id'].nunique()
print(user_details_by_org)

org_id
14      242
28    22170
Name: User_Id, dtype: int64


#### Seems there is substantial representation from organization with org_id = 28

### Get the views per user and group by organization

In [18]:
content_views_df = pd.read_csv("../data/stream_views/0907to0908.csv", encoding = "ISO-8859-1")
print(content_views_df.head())

   USERID  CARDID      CARDTITLE  STREAMID STREAMNAME            TIMESTAMP  \
0    1019    3021  Best in Class       579  About HMD  2017-07-09 10:23:36   
1    1054    3021  Best in Class       579  About HMD  2017-07-09 14:41:25   
2     977    3021  Best in Class       579  About HMD  2017-07-09 13:35:38   
3    1337    3021  Best in Class       579  About HMD  2017-07-10 05:53:05   
4    1346    3021  Best in Class       579  About HMD  2017-07-10 11:23:59   

   MODULEID MODULENAME  ORGANIZATION  
0       473        HMD            28  
1       473        HMD            28  
2       473        HMD            28  
3       473        HMD            28  
4       473        HMD            28  


In [20]:
content_views_per_organization = content_views_df.groupby('ORGANIZATION')["STREAMID"].nunique()
print(content_views_per_organization)

ORGANIZATION
28    136
Name: STREAMID, dtype: int64


### So this file only has streams of organization 28
#### The number of unique streams that has been viewed
#### There are *136* streams in total in the file

In [30]:
unique_num_content_views_by_user = content_views_df.groupby('USERID')["STREAMID"].nunique()
print(unique_num_content_views_by_user)

USERID
950      3
958      4
959     10
961      6
963     21
964      2
977     22
1019     5
1020     1
1021     4
1027    10
1034     4
1035     1
1038     1
1039     2
1044     2
1046     5
1054    22
1058     4
1059    17
1063    20
1064    20
1065    13
1066     1
1067     3
1076     4
1078     5
1079     2
1080     5
1081    10
        ..
2760     2
2761    13
2762     9
2763    19
2764    11
2765     4
2767     5
2768     5
2769     4
2770     7
2771     5
2772     2
2773     8
2774     7
2775     1
2776     2
2777     2
2778     6
2782    13
2783     6
2787     2
2788     1
2789     3
2790     2
2792     5
2794    10
2797     9
2799    19
2800     2
2803     6
Name: STREAMID, Length: 1320, dtype: int64


In [27]:
# Generate a pivot table from the views data
content_views_by_user = pd.pivot_table(content_views_df, index = "USERID", columns = "STREAMID", values = "TIMESTAMP", aggfunc="count")

# fill the NA values with 0 - since there are no views for these contents
content_views_by_user = content_views_by_user.fillna(value = 0)

content_views_by_user.head()

STREAMID  579   580   581   582   583   584   585    586   587   588 ...   \
USERID                                                               ...    
950       0.0   0.0   0.0   0.0  12.0   0.0   0.0    0.0   0.0   0.0 ...    
958       0.0   5.0   0.0   0.0   0.0   0.0   0.0    0.0   0.0  14.0 ...    
959       0.0  31.0  24.0   7.0   0.0   6.0   0.0   32.0   0.0   0.0 ...    
961       0.0  14.0   9.0   0.0   0.0  16.0   0.0   34.0   0.0   0.0 ...    
963       7.0  36.0  21.0  24.0   0.0  28.0  21.0  101.0  51.0  44.0 ...    

STREAMID  786  787  788  789  790  793  795  796  797  798  
USERID                                                      
950       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
958       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
959       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
961       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
963       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 136 columns]

In [29]:
content_views_by_user.to_csv("../data/generated/content_views_by_user_op.csv")

# the 5 quantiles
quantile_values = [0, 0.25, 0.5, 0.75, 1]

# quantiles by users
quantiles_by_users = content_views_by_user.quantile(quantile_values, axis = 0)

# quantiles by streams
quantiles_by_streams = content_views_by_user.quantile(quantile_values, axis = 1)

print(quantiles_by_users)
print(quantiles_by_streams)

STREAMID    579    580    581    582   583    584    585     586     587  \
0.00        0.0    0.0    0.0    0.0   0.0    0.0    0.0     0.0     0.0   
0.25        0.0    0.0    0.0    0.0   0.0    0.0    0.0     0.0     0.0   
0.50        0.0    4.0    2.0    2.0   0.0    1.0    0.0     0.0     0.0   
0.75        3.0   10.0    8.0    7.0   0.0    9.0    6.0    17.0    14.0   
1.00      222.0  433.0  491.0  437.0  12.0  739.0  648.0  7247.0  2222.0   

STREAMID     588 ...    786  787  788  789  790  793  795  796  797  798  
0.00         0.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
0.25         0.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
0.50         0.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
0.75        13.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.00      1707.0 ...   14.0  6.0  6.0  6.0  6.0  8.0  1.0  4.0  2.0  1.0  

[5 rows x 136 columns]
USERID  950   958   959   961    963   964   977   1019  1020  1021  

### Considering the number of streams that each user has viewed

In [32]:
quantiles_by_num_views = unique_num_content_views_by_user.quantile(quantile_values)
print(quantiles_by_num_views)

0.00     1.0
0.25     3.0
0.50     7.0
0.75    16.0
1.00    41.0
Name: STREAMID, dtype: float64


In [ ]:
content_views_by_user_by_card = content_views_by_user